incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSubscribeFactory.java

```diff
@@ -282,52 +346,95 @@ private void onMqttFlush(
             final int reserved = flush.reserved();
 
             assert acknowledge <= sequence;
-            assert sequence >= replySeq;
+            assert sequence >= initialSeq;
+            assert acknowledge >= initialAck;
 
-            replySeq = sequence;
+            initialSeq = sequence;
 
-            assert replyAck <= replySeq;
+            assert initialAck <= initialSeq;
 
             final OctetsFW extension = flush.extension();
             final MqttFlushExFW mqttFlushEx = extension.get(mqttFlushExRO::tryWrap);
 
             assert mqttFlushEx.kind() == MqttFlushExFW.KIND_SUBSCRIBE;
             final MqttSubscribeFlushExFW mqttSubscribeFlushEx = mqttFlushEx.subscribe();
 
-            filters = mqttSubscribeFlushEx.filters();
-            subscriptionIds.clear();
+            currentFilters = mqttSubscribeFlushEx.filters();
+            messagesSubscriptionIds.clear();
+
+            final KafkaFlushExFW kafkaFlushEx = createKafkaFlushEx(currentFilters);
 
+            messages.doKafkaFlush(traceId, authorization, budgetId, reserved, kafkaFlushEx);
 
-            final KafkaFlushExFW kafkaFlushEx =
-                kafkaFlushExRW.wrap(writeBuffer, FlushFW.FIELD_OFFSET_EXTENSION, writeBuffer.capacity())
-                    .typeId(kafkaTypeId)
-                    .merged(m ->
+            if (retainAvailable)
+            {
+                retainedFilters.removeIf(rf -> !currentFilters.anyMatch(f -> f.pattern().asString().equals(rf.filter)));
+                final List<Subscription> newRetainedFilters = new ArrayList<>();
+                currentFilters.forEach(filter ->
+                {
+                    final boolean sendRetained = (filter.flags() & SEND_RETAIN_FLAG) != 0;
+                    final boolean alreadySentRetained =
+                        retainedFilters.stream().anyMatch(rf -> rf.filter.equals(filter.pattern().asString()));
+                    if (sendRetained && !alreadySentRetained)
+                    {
+                        newRetainedFilters.add(new Subscription(
+                            (int) filter.subscriptionId(), filter.pattern().asString(), filter.qos(), filter.flags()));
+                        final boolean rap = (filter.flags() & RETAIN_AS_PUBLISHED_FLAG) != 0;
+                        retainAsPublished.put((int) filter.subscriptionId(), rap);
+                    }
+                });
+
+                if (!newRetainedFilters.isEmpty())
+                {
+                    if (MqttKafkaState.initialOpened(retained.state) && !MqttKafkaState.initialClosed(retained.state))
+                    {
+                        deferredRetainedFiltersQueue.add(newRetainedFilters);
```

Comment:
What problem are we trying to solve by queuing the retained filters?
If we can solve it without introducing buffering state, that would be better. 😄

Reply:
This was a valid concern of yours:
"we might not need to buffer if we initiate the new retained stream for any new filters and defer sending reply window to let it send messages until after the previous retained stream is caught up to live and closed
so there would potentially be more than one retained stream established, but only one allowed to send messages at a time"

So we don't allow more than 1 retained stream to be open. But multiple new retained filter can come in a short amount of time.

Reply:
Ah, by buffer I had meant deferred reaction, not changing internal representation. I wasn't expecting an object data structure.

Reply:
Use the already opened retained stream and change the filters (check if it introduces any problems)

Reply:
this change doesn't seem to have been applied to the code